chain/chain/src/blocks_delay_tracker.rs

```diff
@@ -0,0 +1,204 @@
+        if let Some(block_entry) = self.blocks_in_progress.get_mut(block_hash) {
+            block_entry.orphaned_timestamp = Some(timestamp);
+        } else {
+            debug_assert!(false);
+            warn!(target:"blocks_delay_trakcer", "block {:?} was orphaned but was not marked received", block_hash);
+        }
+    }
+
+    pub fn mark_block_unorphaned(&mut self, block_hash: &CryptoHash, timestamp: Instant) {
+        if let Some(block_entry) = self.blocks_in_progress.get_mut(block_hash) {
+            block_entry.removed_from_orphan_timestamp = Some(timestamp);
+        } else {
+            debug_assert!(false);
+            warn!(target:"blocks_delay_trakcer", "block {:?} was unorphaned but was not marked received", block_hash);
+        }
+    }
+
+    pub fn mark_block_has_missing_chunks(&mut self, block_hash: &CryptoHash, timestamp: Instant) {
+        if let Some(block_entry) = self.blocks_in_progress.get_mut(block_hash) {
+            block_entry.missing_chunks_timestamp = Some(timestamp);
+        } else {
+            debug_assert!(false);
+            warn!(target:"blocks_delay_trakcer", "block {:?} was marked as having missing chunks but was not marked received", block_hash);
+        }
+    }
+
+    pub fn mark_block_completed_missing_chunks(
+        &mut self,
+        block_hash: &CryptoHash,
+        timestamp: Instant,
+    ) {
+        if let Some(block_entry) = self.blocks_in_progress.get_mut(block_hash) {
+            block_entry.removed_from_missing_chunks_timestamp = Some(timestamp);
+        } else {
+            debug_assert!(false);
+            warn!(target:"blocks_delay_trakcer", "block {:?} was marked as having no missing chunks but was not marked received", block_hash);
+        }
+    }
+
+    pub fn mark_chunk_received(&mut self, chunk_hash: &ChunkHash, timestamp: Instant) {
+        self.chunks_in_progress
+            .get_mut(&chunk_hash)
+            .map(|chunk_in_progress| chunk_in_progress.chunk_received.get_or_insert(timestamp));
+    }
+
+    pub fn mark_chunk_requested(
+        &mut self,
+        chunk_hash: &ChunkHash,
+        timestamp: Instant,
+        requestor_block_hash: &CryptoHash,
+    ) {
+        self.chunks_in_progress.entry(chunk_hash.clone()).or_insert_with(|| ChunkInProgress {
+            chunk_requested: timestamp,
+            chunk_received: None,
+            block_hash: requestor_block_hash.clone(),
+        });
+    }
+
+    pub fn finish_block_processing(
+        &mut self,
+        processed_block_hash: &CryptoHash,
+        chunks: &[ChunkHash],
+    ) {
+        self.update_various_block_metrics(processed_block_hash);
+        self.update_block_chunks_requested_metric(processed_block_hash, &chunks);
+        self.update_chunks_metric(&chunks);
+
+        self.blocks_in_progress.remove(&processed_block_hash);
+        for chunk_hash in chunks {
+            self.chunks_in_progress.remove(chunk_hash);
+        }
```

Comment:
This does multiple unnecessary lookups in hash tables. E.g. the block hash is looked up in self.blocks_in_progress three times in this method. Dunno, but I’d do something like:

```rust
let processed_block = self.blocks_in_progress.remove(&block_hash);
self.update_block_metrics(&processed_block);
for chunk_hash in chunks {
    let chunk = self.chunks_in_progress.remove(chunk_hash);
    self.update_chunk_metrics(&processed_block, &chunk);
}
```